lib/internal/modules/esm/resolve.js

```diff
@@ -132,13 +133,34 @@ function emitLegacyIndexDeprecation(url, packageJSONUrl, base, main) {
 
 const realpathCache = new SafeMap();
 
-/**
- * @param {string | URL} url
- * @returns {boolean}
- */
-function fileExists(url) {
-  return internalModuleStat(toNamespacedPath(toPathIfFileURL(url))) === 0;
-}
+const legacyMainResolveExtensions = [
+  '',
+  '.js',
+  '.json',
+  '.node',
+  '/index.js',
+  '/index.json',
+  '/index.node',
+  './index.js',
+  './index.json',
+  './index.node',
+];
+
+const legacyMainResolveReturnType = {
+  // 0-6: when packageConfig.main is defined
+  kResolvedByMain: 0,
+  kResolvedByMainJs: 1,
+  kResolvedByMainJson: 2,
+  kResolvedByMainNode: 3,
+  kResolvedByMainIndexJs: 4,
+  kResolvedByMainIndexJson: 5,
+  kResolvedByMainIndexNode: 6,
+  // 7-9: when packageConfig.main is NOT defined,
+  //      or when the previous case didn't found the file
+  kResolvedByPackageAndJs: 7,
+  kResolvedByPackageAndJson: 8,
+  kResolvedByPackageAndNode: 9,
+};
```

Comment:
It took me a while to realize that these values are the indexes of the `legacyMainResolveExtensions` array. Is it faster to do it this way than a merged data structure like an object or `SafeMap`?

If this is the fastest option, perhaps name this object `legacyMainResolveExtensionsIndexes` to make the connection clearer?